pkg/cableengine/ipsec/ipsec.go

```diff
@@ -95,7 +95,30 @@ func (i *Engine) StartEngine(ignition bool) error {
 		forwardToSubPostroutingRuleSpec := []string{"-j", "SUBMARINER-POSTROUTING"}
 		ipt.AppendUnique("nat", "POSTROUTING", forwardToSubPostroutingRuleSpec...)
 		forwardToSubForwardRuleSpec := []string{"-j", "SUBMARINER-FORWARD"}
-		ipt.AppendUnique("filter", "FORWARD", forwardToSubForwardRuleSpec...)
+		rules, err := ipt.List("filter", "FORWARD")
+		if err != nil {
+			klog.Fatalf("error listing the rules in FORWARD chain: %v", err)
+		}
+
+		appendAt := len(rules) + 1
+		insertAt := appendAt
+		for i, rule := range rules {
+			if rule == "-A FORWARD -j SUBMARINER-FORWARD" {
+				insertAt = -1
+				break
+			} else if rule == "-A FORWARD -j REJECT --reject-with icmp-host-prohibited" {
+				insertAt = i
+				break
+			}
+		}
+
+		if insertAt == appendAt {
+			// Append the rule at the end of the FORWARD Chain.
+			ipt.AppendUnique("filter", "FORWARD", forwardToSubForwardRuleSpec...)
+		} else if insertAt > 0 {
```

Comment:
I don't know if it's possible for the REJECT rule to be at index 0 (I think you've stated previously that there's always or should be some default rule present) but perhaps we should at least log that case with a Warn, eg "A REJECT rule was unexpectedly found at the beginning of the FORWARD chain - the SUBMARINER-FORWARD rule was not added". Or does Insert support index 0?

Reply:
This is a sample output of FORWARD Chain.

	Chain FORWARD (policy DROP 0 packets, 0 bytes)
	num   pkts bytes target     prot opt in     out     source               destination
	1     9638   40M DOCKER-USER  all  --  *      *       0.0.0.0/0            0.0.0.0/0
	2     9638   40M DOCKER-ISOLATION-STAGE-1  all  --  *      *       0.0.0.0/0            0.0.0.0/0
	3     5238   40M ACCEPT     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
	4        0     0 DOCKER     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0
	5     4400  250K ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0
	6        0     0 ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0
	7        0     0 REJECT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            reject-with icmp-host-prohibited


ipt.List() on that host would return the following output.

    iptables_test.go:373: Rule number: 0, rule: -P FORWARD DROP 
    iptables_test.go:373: Rule number: 1, rule: -A FORWARD -j DOCKER-USER 
    iptables_test.go:373: Rule number: 2, rule: -A FORWARD -j DOCKER-ISOLATION-STAGE-1 
    iptables_test.go:373: Rule number: 3, rule: -A FORWARD -o docker0 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT 
    iptables_test.go:373: Rule number: 4, rule: -A FORWARD -o docker0 -j DOCKER 
    iptables_test.go:373: Rule number: 5, rule: -A FORWARD -i docker0 ! -o docker0 -j ACCEPT 
    iptables_test.go:373: Rule number: 6, rule: -A FORWARD -i docker0 -o docker0 -j ACCEPT 
    iptables_test.go:373: Rule number: 7, rule: -A FORWARD -j REJECT --reject-with icmp-host-prohibited 

The zero'th entry which is a default Policy (aka rule number 0) associated with the Chain, cannot have a rule like "REJECT --reject-with icmp-host-prohibited" as shown below.
So, WARN log message is not required. 

[centos@workerone iptables]$ sudo iptables -P FORWARD REJECT --reject-with icmp-host-prohibited                                                                                             
iptables v1.4.21: unknown option "--reject-with"
Try `iptables -h' or 'iptables --help' for more information.